Update chartest

mark-r-anderson opened the pull request:
This PR improves chartest and makes it stricter.

* Switch from checking bad characters to checking allowed characters
  * Way too many potential bad characters
  * Only allow standard ASCII characters
* Add more extensions to check
  * Include documentation and RAT-specific extensions
  * Check many more files than before
* Add quick check of potential non-standard file encoding

@BenLand100 whenever you have the chance, thanks!

mark-r-anderson committed:
Switch to check for allowed characters.

- Switch from checking bad characters to checking allowed characters.
- Add more extensions to check.
- Add more robust check of file encoding.

BenLand100 commented:
While likely still not the most computationally intense RAT test, I note that the change from `line.count(y)` for a specific list of bad characters to a python for loop over every single character in the RAT source code may have a significant performance hit. Maybe not, considering the number of unicode characters previously checked... Is anything substantial gained from this?

It might be better to restrict to ASCII characters by just checking `line.decode('ascii')` doesn't raise an exception, and maybe checking min/max ordinal values to restrict to a range. Lines that fail could potentially be checked for which specific bad characters exist (if we care).

mark-r-anderson committed:
Switch to check range of numeric value of char.

mark-r-anderson commented:
This test is actually only run on the diff relative to master, so the time taken on a typical PR is negligible relative to pretty much any other test (way less than a second per file). You are correct that it does take a performance hit, although it is fairly small relatively speaking because there were quite a few bad characters checked before (plus I think something else in the code is the bottleneck). My tests before submitting the PR showed that when running over the entire source code, the time increase was about 15%.

The reason I did it this way is there is a large number of non-ASCII characters which we feel should not be allowed, and have occasionally made their way in and broken something (e.g., " vs. ” in ratdb files, ” is not ASCII and was not in the disallowed character list), and so checking allowed characters is better. That being said, I like your idea and I have changed it to check the range instead and also added an exception list for outside the range. Sadly, while faster than what I had before, there is basically no performance improvement between this version and the old chartest. Something else must be a bottleneck, but I don’t think it matters much given that this test is already quite fast on most PRs.
